Add script to parse changes from one version to the next and record in documentation

tvhees opened the pull request:
---

**Description of the proposed changes**

Adds a command `yarn build:summarize-changes` which:

- Reads changes to the files in `src/internal/reference/generated` via git
- Extracts changes to actual interfaces (not doc blocks)
- Writes a file to `docs/changelog/${version}.md` with the changes formatted as diffs

**Screenshots (if applicable)**

Example from upcoming release:
![image](https://github.com/user-attachments/assets/680c2d74-6621-4b14-921d-aa48656bf691)

**Other solutions considered (if any)**

This is a quick solution mostly generated by AI. I suspect proper use of `ts-morph` and `diff` libraries would result in a better implementation, but this isn't runtime code and the output is good enough for now.

**Notes to reviewers**

🛈 When you've finished leaving feedback, please add a final comment to the PR tagging the author, letting them know that you have finished leaving feedback

ryanrixxh commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,334 @@
+                 changes.push(`- **Added Interface:** \`${name}\``);
+            }
+        } else if (oldI && !newI) {
+             // Avoid adding empty interfaces (e.g., if only JSDoc was removed)
```
This comment can be moved to the top of the if statement. Its relevant to both

ryanrixxh commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,334 @@
+ */
+function compareInterfaces(oldInterfaces, newInterfaces) {
+    const changes = [];
+    const allNames = new Set([...oldInterfaces.keys(), ...newInterfaces.keys()]);
```
Why a Set if allNames isnt getting compared for differences? Only the nested sets are? Or am I getting confused?

tvhees replied:
The Set is just to avoid duplicate names. This could probably be hardcoded, to be honest, as the root keys are unlikely to be anything but `paths`, `components`, `operations` for both old and new all the time.

ryanrixxh commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,334 @@
+
+
+            // Parse and compare interfaces
+            const oldInterfaces = oldContentStripped
```
`stripAllJsDocs` seems to always return at the very least source code as a string. Why are we checking for falsy here?

tvhees replied:
If a complete file was added (or removed), `getFileContentAtRef` would return null and `stripAllJsDocs` would return `''` which is falsey. So technically this avoids writing a pointless empty source file (I don't think it would error) but yeah, doesn't seem valuable.

ryanrixxh reviewed (changes requested):
I strongly recommend a README for this script. Its really cool, but quite difficult to follow with only JSDocs.

tvhees commented:
@ryanrixxh I've added some more documentation and cleaned up the code a bit. Take another look. A full rewrite with ts-morph was getting very complex and this does the job.

kai-nguyen-aligent commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,363 @@
+            return '';
+        }
+        console.warn(`Warning: Could not get content of ${filePath} at ${gitRef}. Error: ${error} Stderr: ${stderr}`);
+        return '';
```
I think we can utilise something like [simple-git](https://github.com/steveukx/git-js) for this part in the future.

kai-nguyen-aligent commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,363 @@
+    } else {
+        console.log(`File ${filePath} does not exist in working directory (likely deleted or renamed away).`);
+    }
+    return '';
```
I'm not a fan of if-else + return like this. I often return early if the file does not exist :D

tvhees replied:
Have reworked this a bit part of adding simple git. I think we're ok to treat files that didn't exist in a previous ref as empty strings, but have now made it throw an error if there's some other problem instead of just warning.

kai-nguyen-aligent commented on scripts/summarize-changes.js:
```diff
@@ -0,0 +1,363 @@
+                changes.push(`- **Removed Interface:** \`${name}\``);
+            continue;
+             }
+
```
The indentation are off here (and some other part of this file). Is there something wrong with your formatter?

tvhees replied:
It's because formatting is run through eslint, and that's set to ignore js files. I'll run prettier across it.

kai-nguyen-aligent reviewed (dismissed):
Look good to me. I only see few minor format issues but can be address later.

tvhees commented:
@kai-nguyen-aligent @ryanrixxh I have cleaned up the logic here quite a bit - thanks for the impetus to do so. I think the output looks a lot better now too.

![image](https://github.com/user-attachments/assets/61eef3e3-a0f1-43b7-b9fc-fbf7fde105a3)

Ready for re-review!

kai-nguyen-aligent reviewed (approved):
Look good to me. Much cleaner than before 👍🏼
